### PR TITLE
Fix crash in SparseApplyMomentum with a grad of mismatched rank

### LINKED ISSUE #94131: Segmentation Fault in `tf.raw_ops.ResourceSparseApplyMomentum`
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

2.20.0-dev20250516

### Custom code

Yes

### OS platform and distribution

Linux

### Mobile device

_No response_

### Python version

_No response_

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

`tf.raw_ops.ResourceSparseApplyMomentum` causes a segmentation fault when called with mismatched tensor shapes, specifically when the gradient tensor shape doesn't align with the variable dimensions and indices.

Fatal Error: `Check failed: d < dims() (1 vs. 1) in tensorflow/core/framework/tensor_shape.cc:359`

Stack Trace Location:

Primary failure point: `tensorflow::SparseApplyMomentumOp<Eigen::bfloat16, int>::Compute`
Shape validation failure: `tensorflow::TensorShapeBase<tensorflow::TensorShape>::dim_size(int) const`

colab: https://colab.research.google.com/drive/1OZrZ-tWh6m3ojrLX_2vsGIVsuzj1DQEg?usp=sharing

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
import numpy as np

print("TensorFlow version:", tf.__version__)

# Create resource variables
var = tf.Variable(np.random.random((6, 6)).astype(np.float32), dtype=tf.bfloat16)
accum = tf.Variable(np.zeros((6, 6)).astype(np.float32), dtype=tf.bfloat16)

# Create tensors with problematic configurations
lr_tensor = tf.constant(0.0, dtype=tf.bfloat16)
grad_tensor = tf.constant(np.zeros(10), dtype=tf.bfloat16)  # Shape [10] - mismatch
indices_tensor = tf.constant([0, 0, 0, 1, 1, 2, 3, 4], dtype=tf.int32)  # 8 indices
momentum_tensor = tf.constant(0.9, dtype=tf.bfloat16)

# This crashes the process
result = tf.raw_ops.ResourceSparseApplyMomentum(
    var=var.handle,
    accum=accum.handle,
    lr=lr_tensor,
    grad=grad_tensor,
    indices=indices_tensor,
    momentum=momentum_tensor,
    use_locking=True,
    use_nesterov=True
)
```

### Relevant log output

```shell
2025-05-25 19:36:24.885970: F tensorflow/core/framework/tensor_shape.cc:359] Check failed: d < dims() (1 vs. 1)
Aborted (core dumped)
```

### PR DESCRIPTION
**Pull request by GodlyDonuts**
Fixes #94131.

### Problem

`tf.raw_ops.ResourceSparseApplyMomentum` (and `SparseApplyMomentum`) crashes the process when `grad` has fewer dimensions than `var`:

```python
import numpy as np, tensorflow as tf
var = tf.Variable(np.zeros((6, 6), np.float32))
accum = tf.Variable(np.zeros((6, 6), np.float32))
tf.raw_ops.ResourceSparseApplyMomentum(
    var=var.handle, accum=accum.handle,
    lr=0.1, grad=tf.zeros([10]), indices=[0, 1, 2], momentum=0.9)  # crash
```

`SparseApplyMomentumOp::Compute` validates the per-dimension sizes of `grad` against `var` with a loop over `var.dims()`:

```cpp
for (int d = 1; d < var.dims(); d++) {
  OP_REQUIRES(ctx, var.dim_size(d) == grad.dim_size(d), ...);
}
```

but it never checks that `grad` has the same rank as `var` first. When `grad` is lower-rank, `grad.dim_size(d)` reads past `grad`'s rank and aborts the process.

### Fix

Add the same `grad.dims() == var.dims()` guard that the sibling sparse-apply kernels already perform before this loop (e.g. `SparseApplyProximalGradientDescentOp` and `SparseApplyAdagradOp` in the same file). This makes the subsequent `grad.dim_size(d)` accesses well-defined and turns the crash into an `InvalidArgument` error.

Added a regression test (`testResourceSparseApplyMomentumInvalidGradRank`).

### CODE REVIEW

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Hi @GodlyDonuts, thank you for submitting this pull request to resolve the crash in `SparseApplyMomentum` ([#94131](https://github.com/tensorflow/tensorflow/issues/94131)).
  
  I have reviewed the changes, verified they compile, and confirmed that the regression test succeeds inside Google’s testing suite.
  
  While reviewing, we noticed that several other sparse training operators in the same file compile-time validate dimensions but are missing this exact `grad.dims() == var.dims()` rank-check gate and will crash when run with lower-rank gradients:
  * `SparseApplyAdadelta`
  * `SparseApplyProximalGradientDescent`
  * `SparseApplyProximalAdagrad`
  * `SparseApplyRMSProp`
  * `SparseApplyCenteredRMSProp`
  * `SparseApplyKerasMomentum`
  
  Would you be interested in expanding this PR to add the rank verify gate to those sibling operators as well? This would help secure the rest of the sparse apply operators in the codebase.
  
  Let me know if you would like to include updates for those in this PR, or if you prefer to keep this patch scoped to `SparseApplyMomentum` only!

**Comment by GodlyDonuts** [comment]
Thanks! Added the gate to all six. I also hit `SparseApplyAdagradDA` and `SparseApplyFtrl`, they contain the same gap, their loop reads `grad.dim_size(d)` with no rank check either, so I gated those too; happy to split them out if you'd prefer. Folded the coverage into a single parametrized test over all the guarded ops.

**Review by dmiltr3 — APPROVED** [review]

**Review by sakethreddy-27 — APPROVED** [review]

**Comment by GodlyDonuts** [comment]
`feedback/copybara` is red, but it's pointing at an internal run I can't open (cl/936841094), every public check here is green, including `training_ops_test_gpu`. The change only adds the same `grad.dims() == var.dims()` guard the Adagrad ops already carry, so it should only reject the rank-mismatched grad that used to crash, never a valid call.

Could you share what the internal check is flagging? I'd rather fix the real thing than guess at it, my hunch is an internal test pinned to the old behavior for one of these ops, and I can turn a fix around quickly.

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
# PR Contributor Message

The TAP failures in cl/943715489 are indeed due to the changes in this CL.

## Cause of Failure
The PR adds a new import `from tensorflow.python.framework import errors` (translated to `from tensorflow.python.framework import errors` internally) to `training_ops_test.py`.
This import is used in the new regression test `testSparseApplyOpsRejectLowerRankGrad`.

However, the dependency `//tensorflow/python/framework:errors` was not added to all test targets that use `training_ops_test.py` as a source.
While the standard test target `training_ops_test` already includes this dependency (via `TRAINING_TEST_DEPS`), the target `training_ops_mlir_test` (and its derived TPU variants like `dragonfish_training_ops_mlir_test_tpu`) does not include this dependency.

This causes a `PyStrictDepsCheck` failure (missing dependency) when building `training_ops_mlir_test`.

## How to Fix
To fix this issue, you need to add `//tensorflow/python/framework:errors` to the dependencies of `training_ops_mlir_test` in `tensorflow/python/training/BUILD`.

```build
distribute_py_strict_test(
    name = "training_ops_mlir_test",
    srcs = [
        "training_ops_test.py",
    ],
    disable_mlir_bridge = False,
    main = "training_ops_test.py",
    deps = [
        # ... existing dependencies ...
        "//tensorflow/python/framework:errors", # Add this line
    ],
)
```

**Review by dmiltr3 — APPROVED** [review]

**Comment by dmiltr3** [comment]
The change in [_pywrap_tensorflow.def](https://github.com/tensorflow/tensorflow/pull/121730/changes#diff-95fe85770398c25bc0d4d2b96a285dbe30030bc21c8e41b9add279d921883c09) is already in master, so delivering this must be a no-op

**Comment by nithyak0204** [comment]
Hi @GodlyDonuts, Closing this PR as the proposed changes have already been merged into the codebase. Thank you for your contribution!
